assets/js/blocks/attribute-filter/block.js

```diff
@@ -157,39 +99,84 @@ const AttributeFilterBlock = ( {
 	 * Compare intersection of all terms and filtered counts to get a list of options to display.
 	 */
 	useEffect( () => {
+		/**
+		 * Checks if a term slug is in the query state.
+		 */
+		const isTermInQueryState = ( termSlug ) => {
+			if ( ! queryState || ! queryState.attributes ) {
+				return false;
+			}
+			return queryState.attributes.some(
+				( { attribute, slug = [] } ) =>
+					attribute === attributeObject.taxonomy &&
+					slug.includes( termSlug )
+			);
+		};
+
 		if ( attributeTermsLoading || filteredCountsLoading ) {
 			return;
 		}
 
-		const newOptions = [];
+		const newOptions = attributeTerms
+			.map( ( term ) => {
+				const filteredTerm = getFilteredTerm( term.id );
 
-		attributeTerms.forEach( ( term ) => {
-			const filteredTerm = getFilteredTerm( term.id );
-			const isChecked = checked.includes( term.slug );
-			const count = filteredTerm ? filteredTerm.count : null;
+				// If there is no match this term doesn't match the current product collection - only render if checked.
+				if (
+					! filteredTerm &&
+					! checked.includes( term.slug ) &&
+					! isTermInQueryState( term.slug )
+				) {
+					return null;
+				}
 
-			// If there is no match this term doesn't match the current product collection - only render if checked.
-			if ( ! filteredTerm && ! isChecked ) {
-				return;
-			}
+				const count = filteredTerm ? filteredTerm.count : 0;
 
-			newOptions.push( {
-				value: term.slug,
-				name: term.name,
-				label: getLabel( term.name, count ),
-			} );
-		} );
+				return {
+					value: term.slug,
+					name: term.name,
+					label: (
+						<Label
+							name={ term.name }
+							count={ blockAttributes.showCounts ? count : null }
+						/>
+					),
+				};
+			} )
+			.filter( Boolean );
 
 		setDisplayedOptions( newOptions );
 	}, [
+		attributeObject.taxonomy,
 		attributeTerms,
 		attributeTermsLoading,
+		blockAttributes.showCounts,
 		filteredCountsLoading,
 		getFilteredTerm,
-		getLabel,
 		checked,
+		queryState.attributes,
 	] );
 
+	// Track checked STATE changes - if state changes, update the query.
+	useEffect( () => {
+		if ( ! blockAttributes.showFilterButton ) {
+			onSubmit();
+		}
+	}, [ checked, onSubmit ] );
```

Comment:
Does `blockAttributes.showFilterButton` need to be a dependency here?

Reply:
Good point! I think it's not needed because `blockAttributes.showFilterButton` can only change in the editor and `onSubmit` should only be triggered in the frontend.

But that makes me thing that when eluding dependencies we must always explain it in a comment so it doesn't look like a bug. Added in 909a4f9039e69db09b58ab93b75df29692d985a0.